### PR TITLE
Fully typed Slack Search API

### PR DESCRIPTION
**Pull request by leogrim**
@ryanpbrewster

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackAuth.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+import com.keepit.common.crypto.CryptoSupport
+import play.api.libs.json._
+import play.api.libs.functional.syntax._
+import com.kifi.macros.json
+
+import scala.util.{ Failure, Try }
+
+case class SlackAuthScope(value: String)
+object SlackAuthScope {
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackAuth.scala:11]: Nice!

File: kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackAuth.scala [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  userName: SlackUsername,
+  teamId: SlackTeamId,
+  userId: SlackUserId)
+object SlackIdentifyResponse {
+  implicit val reads: Reads[SlackIdentifyResponse] = (
+    (__ \ 'url).read[String] and
+    (__ \ 'team).read[String].map(SlackTeamName(_)) and
+    (__ \ 'user).read[String].map(SlackUsername(_)) and
+    (__ \ 'team_id).read[String].map(SlackTeamId(_)) and
+    (__ \ 'user_id).read[String].map(SlackUserId(_))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackAuth.scala:93]: You added formatters for all of these things right?

File: kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackSearch.scala [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  case class Query(query: String) extends Param("query", Some(query))
+  object Query {
+    def apply(queries: Query*): Query = Query(queries.map(_.query).mkString(" "))
+
+    implicit def fromString(query: String): Query = Query(query)
+    def in(channelName: SlackChannelName) = Query(s"in:${channelName.value}")
+    def from(username: SlackUsername) = Query(s"from:${username.value}")
+    def before(date: LocalDate) = Query(s"before:$date")
+    def after(date: LocalDate) = Query(s"after:$date")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackSearch.scala:21]: The `.toString` method for `LocalDate` does the right thing?

File: kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackSearch.scala [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import com.kifi.macros.json
+
+case class SlackSearchRequest(query: SlackSearchRequest.Query, optional: SlackSearchRequest.Param*)
+
+object SlackSearchRequest {
+  sealed abstract class Param(val name: String, val value: Option[String])
+
+  case class Query(query: String) extends Param("query", Some(query))
+  object Query {
+    def apply(queries: Query*): Query = Query(queries.map(_.query).mkString(" "))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackSearch.scala:15]: We URL-encode the query params, right?

File: kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackSearch.scala (hunk 3 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackSearch.scala:21]: yep

File: kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackSearch.scala (hunk 4 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackSearch.scala:15]: yep

File: kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackAuth.scala (hunk 2 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/models/SlackAuth.scala:93]: fixed

**Pull request merged**